Arithmetic and geometric sequences in semirings (#1475)
UniMath/agda-unimath: src/ring-theory/geometric-sequences-semirings.lagda.md

> # Geomeric sequences in semirings

Reviewer: ```suggestion
# Geometric sequences in semirings
```

Reviewer: Could you add this typo to `codespell-dictionary.txt` please?

Author: > Could you add this typo to `codespell-dictionary.txt` please?

Sure. Is it case-sensitive?

Reviewer: it should be spelled with lowercase. but I fixed it now, so don't worry about it :)

```agda
module ring-theory.geometric-sequences-semirings where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.natural-numbers

open import foundation.action-on-identifications-binary-functions
open import foundation.action-on-identifications-functions
open import foundation.binary-transport
open import foundation.dependent-pair-types
open import foundation.homotopies
open import foundation.identity-types
open import foundation.propositions
open import foundation.sequences
open import foundation.sets
open import foundation.universe-levels

open import group-theory.arithmetic-sequences-semigroups

open import ring-theory.powers-of-elements-semirings
open import ring-theory.semirings
```

</details>

## Ideas

An
{{#concept "geometric sequence" Disambiguation="in a semiring" Agda=geometric-sequence-Semiring}}
in a [semiring](ring-theory.semirings.md) is an
[arithmetic sequence](group-theory.arithmetic-sequences-semigroups.md) in the
semiring multiplicative [semigroup](group-theory.semigroups.md).

> These are the sequences `n ↦ a * r ⁿ` for some elements `a r` in the semiring.

Reviewer: ```suggestion
These are sequences of the form `n ↦ a * rⁿ`, for elements `a`, `r` in the semiring.
```

  initial-term-geometric-sequence-Semiring =
    initial-term-arithmetic-sequence-Semigroup
      ( multiplicative-semigroup-Semiring R)
      ( u)
```

### The standard geometric sequences in a semiring

The standard geometric sequence with initial term `a` and common factor `r` is
the sequence `u` defined by:

- `u₀ = a`
- `uₙ₊₁ = uₙ * r`

```agda
module _
  {l : Level} (R : Semiring l) (a r : type-Semiring R)
  where

  standard-geometric-sequence-Semiring : geometric-sequence-Semiring R
  standard-geometric-sequence-Semiring =
    standard-arithmetic-sequence-Semigroup
      ( multiplicative-semigroup-Semiring R)
      ( a)
      ( r)

  seq-standard-geometric-sequence-Semiring : ℕ → type-Semiring R
  seq-standard-geometric-sequence-Semiring =
    seq-geometric-sequence-Semiring R
      standard-geometric-sequence-Semiring

  is-geometric-standard-geometric-sequence-Semiring :
    is-arithmetic-sequence-Semigroup
      ( multiplicative-semigroup-Semiring R)
      ( seq-standard-geometric-sequence-Semiring)
  is-geometric-standard-geometric-sequence-Semiring =
    is-geometric-seq-geometric-sequence-Semiring R
      standard-geometric-sequence-Semiring
```

> ### The geometric sequences `n ↦ a * r ⁿ`

Reviewer: ```suggestion
### The geometric sequences `n ↦ a * rⁿ`
```

  geometric-mul-pow-nat-Semiring : geometric-sequence-Semiring R
  geometric-mul-pow-nat-Semiring =
    ( mul-pow-nat-Semiring , is-geometric-mul-pow-nat-Semiring)

  initial-term-mul-pow-nat-Semiring : type-Semiring R
  initial-term-mul-pow-nat-Semiring =
    initial-term-geometric-sequence-Semiring
      ( R)
      ( geometric-mul-pow-nat-Semiring)

  eq-initial-term-mul-pow-nat-Semiring :
    initial-term-mul-pow-nat-Semiring ＝ a
  eq-initial-term-mul-pow-nat-Semiring =
    right-unit-law-mul-Semiring R a
```

## Properties

### Any geometric sequence in a semiring is homotopic to a standard geometric sequence

```agda
module _
  {l : Level} (R : Semiring l)
  (u : geometric-sequence-Semiring R)
  where

  htpy-seq-standard-geometric-sequence-Semiring :
    ( seq-geometric-sequence-Semiring R
      ( standard-geometric-sequence-Semiring R
        ( initial-term-geometric-sequence-Semiring R u)
        ( common-ratio-geometric-sequence-Semiring R u))) ~
    ( seq-geometric-sequence-Semiring R u)
  htpy-seq-standard-geometric-sequence-Semiring =
    htpy-seq-standard-arithmetic-sequence-Semigroup
      ( multiplicative-semigroup-Semiring R)
      ( u)
```

> ### The nth term of an geometric sequence with initial term `a` and common ratio `r` is `a * r ⁿ`

Reviewer: ```suggestion
### The nth term of an geometric sequence with initial term `a` and common ratio `r` is `a * rⁿ`
```
